File: icapserver.pl
#!/usr/bin/perl

use Socket;
use POSIX;

sub icap_session
{
	local(*SHA) = @_;
	local($method, $icap_url);
	local($enc_null_body, $enc_req_hdr);

	$_ = <SHA>;
	tr/\r\n//d;
	print "LINE: [$_]\n";
	if(m|^(\S+) (\S+) ICAP/1.0|){
		$method = $1;
		$icap_url = $2;
	}else{
		die "Invalid request line\n";
	}
	while(<SHA>){
		tr/\r\n//d;
		print "LINE: [$_]\n";
		if(/^$/){print "AAA\n";last;}
		if(/^Encapsulated:/){
			if(/(^|[^\w])req-hdr=(\d+)/){$enc_req_hdr = $2;}
			if(/(^|[^\w])null-body=(\d+)/){$enc_null_body = $2;}
		}
	}
	if($enc_null_body){
		$n = read(SHA, $buf_body, $enc_null_body);
		print "BODY: [$buf_body]\n";	
	}
	if($method eq "OPTIONS"){
		print SHA "ICAP/1.0 200 OK\r\n";
		print SHA "Encapsulated: null-body=0\r\n";
		print SHA "Connection: close\r\n";
		print SHA "Methods: REQMOD\r\n";
		print SHA "ISTag: 1111\r\n";
		print SHA "Transfer-Complete: *\r\n";
		print SHA "Allow: 204\r\n";
		print SHA "\r\n";
	}elsif($method eq "REQMOD"){
		$buf_body =~ /^(\S+) (\S+)/;
		$url = $2;
		print "URL: $url\n";
		if($url =~ /block/){
			print "### BLOCK###\n";
if(0){
			local($respbuf) = <<_EOT_;
ICAP/1.0 200 OK
Date: Mon, 10 Jan 2000  09:55:21 GMT
Server: ICAP-Server-Software/1.0
Connection: close
ISTag: "W3E4R7U9-L2E4-2"
Encapsulated: res-hdr=0, res-body=213

HTTP/1.1 403 Forbidden
Date: Wed, 08 Nov 2000 16:02:10 GMT
Server: Apache/1.3.12 (Unix)
Last-Modified: Thu, 02 Nov 2000 13:51:37 GMT
ETag: "63600-1989-3a017169"
Content-Length: 58
Content-Type: text/html

3a
Sorry, you are not allowed to access that naughty content.
0

_EOT_
			$respbuf =~ tr/\r//d;
			$respbuf =~ s/\n/\r\n/g;
			print "RESPBUF: [$respbuf]\n";
			print SHA $respbuf;
}
if(1){
			local($http_body) = "<h1>This is body</h1>\r\n";
			# local($http_resp_header) = "HTTP/1.1 403 OK\r\nContent-Type: text/html\r\nContent-Length: " . length($http_body) . "\r\n\r\n";
			local($http_resp_header) = "HTTP/1.0 200 OK\r\nContent-Type: text/html\r\nContent-Length: " . length($http_body) . "\r\n\r\n";
			local($resp_body) = $http_resp_header . sprintf("%x", length($http_body)) . "\r\n" . $http_body . "\r\n0\r\n\r\n";
			print "----------\n";
			print SHA "ICAP/1.0 200 OK\r\n";
			print "ICAP/1.0 200 OK\r\n";
			print SHA "Encapsulated: res-hdr=0, res-body=" . (length($http_resp_header)) . "\r\n";
			print "Encapsulated: res-hdr=0, res-body=" . (length($http_resp_header)) . "\r\n";
			print SHA "ISTag: 2222\r\n";
			print "ISTag: 2222\r\n";
			print SHA "Connection: close\r\n";
			print "Connection: close\r\n";
			print SHA "\r\n";
			print "\r\n";
			print SHA $resp_body;
			print $resp_body;
			print "----------\n";
}
		}else{
			print SHA "ICAP/1.0 204 OK\r\n";
			print SHA "Encapsulated: null-body=0\r\n";
			print SHA "ISTag: 2222\r\n";
			print SHA "\r\n";
		}
	}
}

$port = ($ARGV[0] || 1344);
print "LISTEN PORT: $port ...\n";
socket(SH, PF_INET, SOCK_STREAM, 0) || die "cannot open socket: $!";
setsockopt(SH, SOL_SOCKET, SO_REUSEADDR, 1) || die "setsockopt: $!";
bind(SH, sockaddr_in($port, INADDR_ANY)) || die "cannot bind to me: $!";
listen(SH, SOMAXCONN) || die "cannot listen socket: $!";
while(1){
	local(*SHA);
	accept(SHA, SH) || die "cannot accept socket: $!";
	my($rport,$raddr) = unpack_sockaddr_in(getpeername(SHA));
	print "ACCEPT From: ",inet_ntoa($raddr),":",$rport,"\n";
	select((select(SHA),$|=1)[0]);
	&icap_session(*SHA);
	print "CLOSED\r\n";
	close(SHA);
}

File: icapserver.c
/************************************************************
  Sample test icap server

	- Block access that URL contains "icapblock".
************************************************************/



#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <arpa/inet.h>

const char icap_server_name[] = "ICAP-TEST/1.0";

int icap_session(int as)
{
	FILE *fp_r=NULL, *fp_w=NULL;
	char icap_method[100]="";
	char client_ip[100]="";
	int func_ret = -1;
	char buf[1000];
	char *p, *p2;

	fp_r = fdopen(dup(as), "rb");
	if(fp_r==NULL){perror("fp_r"); exit(1);}
	fp_w = fdopen(dup(as), "wb");
	if(fp_w==NULL){perror("fp_w"); exit(1);}

	/*** get ICAP request line ***/
	if(fgets(buf, sizeof(buf), fp_r)==NULL){goto endlabel;}
	p = strchr(buf, ' ');
	if(p==NULL || p-buf >= sizeof(icap_method)){goto endlabel;}
	strncpy(icap_method, buf, p-buf);
	icap_method[p-buf] = '\0';

	/*** get ICAP request header ***/
	while(fgets(buf, sizeof(buf), fp_r)){
		p = strpbrk(buf, "\r\n");
		if(p){*p='\0';}
		if(buf[0]=='\0'){break;}
		if(strncasecmp(buf, "X-Client-IP: ", 13)==0 && strlen(buf+13)<sizeof(client_ip)){
			strcpy(client_ip, buf+13);
		}
	}

	printf("icap_method: [%s]\n", icap_method);

	if(strcmp(icap_method, "OPTIONS")==0){
		fprintf(fp_w, "ICAP/1.0 200 OK\r\n");
		fprintf(fp_w, "Connection: close\r\n");
		fprintf(fp_w, "Encapsulated: null-body=0\r\n");
		fprintf(fp_w, "Methods: REQMOD\r\n");
		fprintf(fp_w, "ISTag: %d\r\n", time(NULL));
		fprintf(fp_w, "Transfer-Complete: *\r\n");
		fprintf(fp_w, "Service: %s\r\n", icap_server_name);
		fprintf(fp_w, "Allow: 204\r\n");
		fprintf(fp_w, "\r\n");
	}else if(strcmp(icap_method, "REQMOD")==0){
		char http_method[100]="";
		char http_url[1000]="";
	
		/*** get HTTP request line ***/
		if(fgets(buf, sizeof(buf), fp_r)==NULL){goto endlabel;}
		p = strchr(buf, ' ');
		if(p==NULL || p-buf >= sizeof(http_method)){goto endlabel;}
		strncpy(http_method, buf, p-buf);
		http_method[p-buf] = '\0';

		p2 = p + 1;while(isspace(*p2)){p2++;}
		p = strchr(p2, ' ');
		if(p==NULL || p-p2 >= sizeof(http_url)){goto endlabel;}
		strncpy(http_url, p2, p-p2);
		http_method[p-p2] = '\0';

		printf("REQMOD: method=[%s], url=[%s], client_ip=[%s]\n", http_method, http_url, client_ip);

		if(strstr(http_url, "icapblock")){
			char http_resp_body[] = "<h1>Blocked</h1>";
			char http_resp_header[1000];

			snprintf(http_resp_header, sizeof(http_resp_header),
				"HTTP/1.0 200 OK\r\n"
				"Content-Type: text/html\r\n"
				"Content-Length: %d\r\n"
				"\r\n"
				, strlen(http_resp_body)
			);
			fprintf(fp_w, "ICAP/1.0 200 OK\r\n");
			fprintf(fp_w, "Encapsulated: res-hdr=0, res-body=%d\r\n", strlen(http_resp_header));
			fprintf(fp_w, "ISTag: %d\r\n", time(NULL));
			fprintf(fp_w, "Server: %s\r\n", icap_server_name);
			fprintf(fp_w, "Connection: close\r\n");
			fprintf(fp_w, "\r\n");
			fputs(http_resp_header, fp_w);
			fprintf(fp_w, "%x\r\n", strlen(http_resp_body));
			fputs(http_resp_body, fp_w);
			fprintf(fp_w, "\r\n0\r\n");
		}else{
			fprintf(fp_w, "ICAP/1.0 204 No Content\r\n");
			fprintf(fp_w, "Encapsulated: null-body=0\r\n");
			fprintf(fp_w, "ISTag: %d\r\n", time(NULL));
			fprintf(fp_w, "Server: %s\r\n", icap_server_name);
			fprintf(fp_w, "Connection: close\r\n");
			fprintf(fp_w, "\r\n");
		}
	}else{
		fprintf(fp_w, "ICAP/1.0 405 Method not allowed for service\r\n");
		fprintf(fp_w, "Encapsulated: null-body=0\r\n");
		fprintf(fp_w, "ISTag: %d\r\n", time(NULL));
		fprintf(fp_w, "Server: %s\r\n", icap_server_name);
		fprintf(fp_w, "Connection: close\r\n");
		fprintf(fp_w, "\r\n");
	}

	func_ret = 0;
endlabel:
	if(fp_r){fclose(fp_r);}
	if(fp_w){fclose(fp_w);}
	return func_ret;
}

int main(int argc, char *argv[])
{
	int listen_port = 1344;
	char listen_addrstr[100] = "127.0.0.1";
	int s;
	struct sockaddr_in sin;
	int val, ret;

	s = socket(AF_INET, SOCK_STREAM, 0);
	if(s==-1){perror("socket:"); exit(1);}

	val = 1;
	ret = setsockopt(s, SOL_SOCKET, SO_REUSEADDR, &val, sizeof(val));
	if(ret==-1){perror("setsockopt(SO_REUSEADDR)"); exit(1);}

	sin.sin_family = AF_INET;
	sin.sin_port = htons(listen_port);
	sin.sin_addr.s_addr = inet_addr(listen_addrstr);
	ret = bind(s, (struct sockaddr*)&sin, sizeof(sin));
	if(ret==-1){perror("bind:"); exit(1);}

	ret = listen(s, 5);
	if(ret==-1){perror("listen:"); exit(1);}

	printf("listening %s:%d...\n", listen_addrstr, listen_port);

	while(1){
		struct sockaddr_in from;
		int fromlen = sizeof(from);
		int as;

		as = accept(s, (struct sockaddr*)&from, &fromlen);
		if(as==-1){perror("accept:"); exit(1);}

		printf("accepted\n");

		icap_session(as);

		close(as);
	}
	return 0;
}
